x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/normalize/NormalizePipelineAggregator.java

```diff
@@ -0,0 +1,78 @@
+import org.elasticsearch.search.aggregations.InternalMultiBucketAggregation;
+import org.elasticsearch.search.aggregations.bucket.MultiBucketsAggregation.Bucket;
+import org.elasticsearch.search.aggregations.bucket.histogram.HistogramFactory;
+import org.elasticsearch.search.aggregations.pipeline.BucketHelpers.GapPolicy;
+import org.elasticsearch.search.aggregations.pipeline.InternalSimpleValue;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static org.elasticsearch.search.aggregations.pipeline.BucketHelpers.resolveBucketValue;
+
+public class NormalizePipelineAggregator extends PipelineAggregator {
+    private final DocValueFormat formatter;
+    private final Function<List<Double>, NormalizePipelineNormalizer> normalizerSupplier;
+
+    NormalizePipelineAggregator(String name, String[] bucketsPaths, DocValueFormat formatter,
+                                Function<List<Double>, NormalizePipelineNormalizer> normalizerSupplier,
+                                Map<String, Object> metadata) {
+        super(name, bucketsPaths, metadata);
+        this.formatter = formatter;
+        this.normalizerSupplier = normalizerSupplier;
+    }
+
+    @Override
+    public InternalAggregation reduce(InternalAggregation aggregation, ReduceContext reduceContext) {
+        InternalMultiBucketAggregation<? extends InternalMultiBucketAggregation, ? extends InternalMultiBucketAggregation.InternalBucket>
+                histo = (InternalMultiBucketAggregation<? extends InternalMultiBucketAggregation, ? extends
+                InternalMultiBucketAggregation.InternalBucket>) aggregation;
+        List<? extends InternalMultiBucketAggregation.InternalBucket> buckets = histo.getBuckets();
+        HistogramFactory factory = (HistogramFactory) histo;
+        List<Bucket> newBuckets = new ArrayList<>(buckets.size());
+
+        List<Double> values = buckets.stream().map(bucket -> resolveBucketValue(histo, bucket, bucketsPaths()[0], GapPolicy.SKIP))
+            .collect(Collectors.toList());
+
+        NormalizePipelineNormalizer normalizer = normalizerSupplier.apply(values);
+
+        for (int i = 0; i < buckets.size(); i++) {
+            InternalMultiBucketAggregation.InternalBucket bucket = buckets.get(i);
+            Double thisBucketValue = values.get(i);
+
+            final double normalizedBucketValue;
+
+            // Only account for finite values
+            if (thisBucketValue.isNaN()) {
+                normalizedBucketValue = Double.NaN;
+            } else {
+                normalizedBucketValue = normalizer.normalize(thisBucketValue);
+            }
+
+            List<InternalAggregation> aggs = StreamSupport.stream(bucket.getAggregations().spliterator(), false)
```

Comment:
`bucket.getAggregations().copyResults()` does this without so much boiler plate.

Reply:
unfortunately, that method does not work in this context. I think a more dedicated cleanup for this boilerplate can be tackled outside of this PR

Reply:
:+1: